coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -1184,16 +1193,25 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
 				}
 				go func() {
 					for _, wait := range updates {
-						// Wait for the next potential timeout to occur. Note that we
-						// can't listen on the context here because we will hang around
-						// after this function has returned. The s also doesn't
-						// have a shutdown signal we can listen to.
-						<-wait
-						if err := s.Pubsub.Publish(codersdk.WorkspaceNotifyChannel(workspaceBuild.WorkspaceID), []byte{}); err != nil {
-							s.Logger.Error(ctx, "workspace notification after agent timeout failed",
+						select {
+						case <-s.lifecycleCtx.Done():
+							// If the server is shutting down, we don't want to wait around.
+							s.Logger.Warn(context.Background(), "stopping notifications due to server shutdown",
 								slog.F("workspace_build_id", workspaceBuild.ID),
 								slog.Error(err),
 							)
+							return
+						case <-wait:
+							// Wait for the next potential timeout to occur. Note that we
+							// can't listen on the context here because we will hang around
+							// after this function has returned. The s also doesn't
+							// have a shutdown signal we can listen to.
```

Comment:
this comment is outdated --- we DO have a context that acts as a shutdown signal for s.